main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -186,54 +178,26 @@ object Lexer {
     * Note: If the lexer has arrived at EOF advance will continuously return EOF without advancing.
     * This is a design choice to avoid returning an Option[Char], which would be doable but tedious to work with.
     */
-  private def advance()(implicit s: State): Char = {
-    if (s.current.offset >= s.src.data.length) {
-      return '\u0000'
-    }
-
-    val c = s.src.data(s.current.offset)
-    if (c == '\n') {
-      s.prevWidth = s.current.column
-      s.current.line += 1
-      s.current.column = 0
-    } else {
-      s.current.column += 1
-    }
-    s.current.offset += 1
-    c
-  }
+  private def advance()(implicit s: State): Char =
+    s.sc.advance()
 
   /**
     * Peeks the previous character that state was on if available.
     */
-  private def previous()(implicit s: State): Option[Char] = {
-    if (s.current.offset == 0) {
-      None
-    } else {
-      Some(s.src.data(s.current.offset - 1))
-    }
-  }
+  private def previous()(implicit s: State): Option[Char] =
+    s.sc.previous
 
   /**
     * Peeks the character before the previous that state was on if available.
     */
-  private def previousPrevious()(implicit s: State): Option[Char] = {
-    if (s.current.offset <= 1) {
-      None
-    } else {
-      Some(s.src.data(s.current.offset - 2))
-    }
-  }
+  private def previousPrevious()(implicit s: State): Option[Char] =
+    s.sc.previousPrevious
```

Comment:
Is this good design? (same comments apply elsewhere)

Reply:
I think it previousPrevious was used for performance? you can skip one of the bound checks if when you know the index I negative. but ill delete this double one at least